File: bada_ruby/lib/bada/quantum/bell.rb
# frozen_string_literal: true

module Bada
  module Quantum
    # Bell's experiment: the CHSH inequality and its quantum violation — the
    # certificate of "spooky action at a distance" (不気味な遠隔作用) between an
    # entangled *pair*.
    #
    #   * A local hidden-variable (classical) world obeys  |S| <= 2.
    #   * Quantum mechanics for a singlet pair predicts  E(a,b) = -cos(a-b),
    #     which reaches the Tsirelson bound  |S| = 2*sqrt(2) ~= 2.828.
    #
    # The gap between the two is not solved by any radical/algebraic completion
    # of a local model — it is genuinely non-local. We expose both the exact
    # quantum prediction and a Monte-Carlo run of a local model, so the
    # violation is shown *statistically* (確率・統計).
    module Bell
      module_function

      # Standard optimal CHSH settings (radians).
      A  = 0.0
      A2 = Math::PI / 2.0
      B  = Math::PI / 4.0
      B2 = 3.0 * Math::PI / 4.0

      TSIRELSON = 2.0 * Math.sqrt(2.0)

      # Quantum correlation of the singlet pair at analyzer angles a, b.
      def e_quantum(a, b)
        -Math.cos(a - b)
      end

      # CHSH combination  S = E(a,b) - E(a,b') + E(a',b) + E(a',b').
      def chsh_quantum(a = A, a2 = A2, b = B, b2 = B2)
        e_quantum(a, b) - e_quantum(a, b2) + e_quantum(a2, b) + e_quantum(a2, b2)
      end

      # One shared-hidden-variable trial of a *local* model. A common lambda is
      # drawn; each side answers with only its own analyzer angle and lambda —
      # no communication. This is the best deterministic local strategy and it
      # cannot beat |S| <= 2.
      def local_outcome(angle, lambda)
        Math.cos(angle - lambda) >= 0 ? +1 : -1
      end

      # Monte-Carlo estimate of a local correlation E(a,b) over N trials.
      def e_local(a, b, trials, rng)
        sum = 0
        trials.times do
          lam = rng.rand * 2 * Math::PI
          sum += local_outcome(a, lam) * local_outcome(b, lam)
        end
        sum.to_f / trials
      end

      def chsh_local(trials: 20_000, rng: Random.new(1))
        eab  = e_local(A,  B,  trials, rng)
        eab2 = e_local(A,  B2, trials, rng)
        ea2b = e_local(A2, B,  trials, rng)
        ea2b2 = e_local(A2, B2, trials, rng)
        eab - eab2 + ea2b + ea2b2
      end

      # The full experiment result: the classical bound, the sampled local S,
      # the exact quantum S, the Tsirelson ceiling, and whether spooky action is
      # certified (quantum S beats the classical bound by more than the sampled
      # local model's spread).
      def experiment(trials: 20_000, seed: 1)
        rng = Random.new(seed)
        s_local = chsh_local(trials: trials, rng: rng)
        s_quantum = chsh_quantum
        {
          classical_bound: 2.0,
          local_chsh: s_local,
          quantum_chsh: s_quantum,
          tsirelson: TSIRELSON,
          violates_local_realism: s_quantum.abs > 2.0 + 1e-9,
          spooky_action_certified: s_quantum.abs > s_local.abs + 0.2,
          margin: s_quantum.abs - 2.0
        }
      end
    end
  end
end

File: apps/core/src/main/java/bada/quantum/Bell.java
package bada.quantum;

import java.util.Random;

/**
 * Bell's experiment: CHSH inequality and its quantum violation — the
 * certificate of spooky action at a distance (不気味な遠隔作用) between a pair.
 * Classical local hidden-variable worlds obey |S| <= 2; the singlet reaches the
 * Tsirelson bound |S| = 2*sqrt(2).
 */
public final class Bell {
    public static final double A  = 0.0;
    public static final double A2 = Math.PI / 2.0;
    public static final double B  = Math.PI / 4.0;
    public static final double B2 = 3.0 * Math.PI / 4.0;
    public static final double TSIRELSON = 2.0 * Math.sqrt(2.0);

    public static double eQuantum(double a, double b) { return -Math.cos(a - b); }

    public static double chshQuantum() {
        return eQuantum(A, B) - eQuantum(A, B2) + eQuantum(A2, B) + eQuantum(A2, B2);
    }

    static int localOutcome(double angle, double lambda) {
        return Math.cos(angle - lambda) >= 0 ? 1 : -1;
    }

    static double eLocal(double a, double b, int trials, Random rng) {
        long sum = 0;
        for (int i = 0; i < trials; i++) {
            double lam = rng.nextDouble() * 2 * Math.PI;
            sum += (long) localOutcome(a, lam) * localOutcome(b, lam);
        }
        return (double) sum / trials;
    }

    public static double chshLocal(int trials, Random rng) {
        double eab  = eLocal(A,  B,  trials, rng);
        double eab2 = eLocal(A,  B2, trials, rng);
        double ea2b = eLocal(A2, B,  trials, rng);
        double ea2b2 = eLocal(A2, B2, trials, rng);
        return eab - eab2 + ea2b + ea2b2;
    }

    /** Result of the Bell experiment. */
    public static final class Result {
        public final double classicalBound = 2.0;
        public double localChsh, quantumChsh, tsirelson = TSIRELSON, margin;
        public boolean violatesLocalRealism, spookyActionCertified;
    }

    public static Result experiment(int trials, long seed) {
        Random rng = new Random(seed);
        Result r = new Result();
        r.localChsh = chshLocal(trials, rng);
        r.quantumChsh = chshQuantum();
        r.violatesLocalRealism = Math.abs(r.quantumChsh) > 2.0 + 1e-9;
        r.spookyActionCertified = Math.abs(r.quantumChsh) > Math.abs(r.localChsh) + 0.2;
        r.margin = Math.abs(r.quantumChsh) - 2.0;
        return r;
    }
}
